ql/src/java/org/apache/hadoop/hive/ql/optimizer/metainfo/annotation/OpTraitsRulesProcFactory.java

```diff
@@ -158,8 +158,57 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
         }
       }
 
-      listBucketCols.add(bucketCols);
-      OpTraits opTraits = new OpTraits(listBucketCols, numBuckets,
+      final List<List<String>> listBucketCols = new ArrayList<>();
+      final List<CustomBucketFunction> bucketFunctions = new ArrayList<>();
+      final int numBuckets;
+      if (parentOpTraits == null || !parentOpTraits.hasCustomBucketFunction()) {
+        // No CustomBucketFunctions
+        listBucketCols.add(bucketCols);
+        bucketFunctions.add(null);
+        numBuckets = parentOpTraits == null ? -1 : parentOpTraits.getNumBuckets();
+      } else if (parentOpTraits.getCustomBucketFunctions().size() > 1) {
+        // We don't know how to merge multiple custom bucket functions. Reset bucket attributes
+        Preconditions.checkState(parentOpTraits.getBucketColNames().size() > 1);
+        listBucketCols.add(Collections.emptyList());
+        bucketFunctions.add(null);
+        numBuckets = -1;
+      } else {
+        Preconditions.checkState(parentOpTraits.getBucketColNames().size() == 1);
+        Preconditions.checkState(parentOpTraits.getCustomBucketFunctions().size() == 1);
```

Comment:
why do we need this check? `1` is the only possible value here

Reply:
I added some assertions because of the complexity of this class. But I also think it is obvious enough that this condition should be satisfied. I removed the validation
https://github.com/apache/hive/pull/5409/commits/c6aa9d2d0a81ff2e8834e1350a18405e593fe212